fix(redirect-parser): handle merge of large redirect files

JGAntunes opened the pull request:
🎉 Thanks for submitting a pull request! 🎉

#### Summary

Fixes https://github.com/netlify/pillar-workflow/issues/897

<!--
Explain the **motivation** for making this change. What existing problem does the pull request solve and how?
-->

---

For us to review and ship your PR efficiently, please perform the following steps:

- [ ] Open a [bug/issue](https://github.com/netlify/build/issues/new/choose) before writing your code 🧑‍💻. This ensures
      we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or
      something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [ ] Read the [contribution guidelines](../CONTRIBUTING.md) 📖. This ensures your code follows our style guide and
      passes our tests.
- [ ] Update or add tests (if any source code was changed or added) 🧪
- [ ] Update or add documentation (if features were changed or added) 📝
- [ ] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**

akahn commented on packages/redirect-parser/src/merge.ts:
```diff
@@ -10,14 +8,18 @@ import { splitResults } from './results.js'
-const isUniqueRedirect = function (redirect, index, redirects) {
-  return !redirects.slice(index + 1).some((otherRedirect) => isDeepStrictEqual(redirect, otherRedirect))
+// The latest duplicate value is the one kept, hence why we need to reverse the
+// arrays in order to keep said logic.
```
If the lines are identical, why does it matter which one is kept?

JGAntunes replied:
The lines are identical however there might be other redirects with the same `from` property, changing the order of the duplicate we keep now could effectively tamper with that logic. A practical example, given the non-deduped array:

```js
[
  {
    from: "this-is-not-unique"
    to: "this-too-is-not-unique"
  },
  {
    from: "this-is-not-unique"
    to: "this-IS-UNIQUE"
  },
  {
    from: "this-is-not-unique"
    to: "this-too-is-not-unique"
  },
]
```

Assuming we wouldn't keep said logic and effectively keep the first duplicate, this would render:

```js
[
  {
    from: "this-is-not-unique"
    to: "this-too-is-not-unique"
  },
  {
    from: "this-is-not-unique"
    to: "this-IS-UNIQUE"
  },
]
```

Now, according to our docs - https://docs.netlify.com/routing/redirects/#rule-processing-order -  the end result for this case given that:
>  "The redirects engine will process the first matching rule it finds, reading from top to bottom. Rules in the _redirects file are always processed first, followed by rules in the Netlify configuration file."
The redirect that would end up being deployed would be the first one `to` `"this-too-is-not-unique"`.

The current implementation however would render a different scenario, `to` `"this-IS-UNIQUE"` would be deployed.

So tl;dr changing that behaviour would be a breaking change.

eduardoboucas commented on packages/redirect-parser/src/merge.ts:
```diff
@@ -10,14 +8,18 @@ import { splitResults } from './results.js'
+  const strRedirects: [string] = redirects.map((v) => JSON.stringify(v)).reverse()
+  // Set takes care of creating a set of non duplicate values and preserves order
+  const set = new Set(strRedirects)
+  return [...set].map((v) => JSON.parse(v)).reverse()
```
I think you could avoid parsing the objects back if you did something like:

```ts
const uniqueRedirects = new Set<string>()
const duplicateIndexes = new Set<number>()

redirects.forEach((v, i) => {
  const str = JSON.stringify(v)

  if (uniqueRedirects.has(str)) {
    duplicateIndexes.add(i)
  } else {
    uniqueRedirects.add(str)
  }
})

const mergedRedirects = redirects.filter((_, i) => !duplicateIndexes.has(i))

return mergedRedirects
```

Not sure how much difference this will make performance-wise, but maybe worth a shot?

JGAntunes replied:
Ah sorry, only saw this now 🤦 after doing something similar for the haders - https://github.com/netlify/build/pull/4653 - where I was forced to go with a different approach I'm thinking it might make sense to use something similar here. Let me know what you think 👍 But I'll run some tests on Monday and that way we'll have more data to compare.

JGAntunes replied:
Ran some more tests locally using a version of:
- https://github.com/netlify/build/blob/4f11454785041ddadf7673ab5ac87034a527cafc/packages/headers-parser/src/merge.ts#L24-L49

Which I believe is similar to what you're describing in there ☝️ 

Used a 200k+ redirects file in order to better highlight the differences.

Using the current `JSON.parse()` approach:
```
bash-5.1$ /usr/bin/time -lh node bench-redirects.js
Generated 45001 redirects
	3.56s real		3.98s user		0.34s sys
           677801984  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
              193561  page reclaims
                   0  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
                   0  voluntary context switches
                7448  involuntary context switches
         62445986557  instructions retired
         13670960738  cycles elapsed
           658956032  peak memory footprint
```

With the mentioned approach (not using JSON.parse):
```
Generated 45001 redirects
	3.48s real		3.78s user		0.27s sys
           452116480  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
              138161  page reclaims
                   0  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
                   0  voluntary context switches
                7410  involuntary context switches
         59548212541  instructions retired
         12775129961  cycles elapsed
           431560128  peak memory footprint
```

The key difference is the memory usage impact (~200MB), so I'm going to repurpose that method to use her too.

JGAntunes commented:
> by using JSON.stringify is there any risk of not detecting duplicates due to key ordering?

That's a great question actually. As per MDN's `JSON.stringify()` docs, its implementation is stable, so keys will have order guarantee:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

> Only [enumerable own properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) are visited. This means [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map), [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set), etc. will become "{}". You can use the [replacer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#the_replacer_parameter) parameter to serialize them to something more useful. Properties are visited using the same algorithm as [Object.keys()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys), which has a well-defined order and is stable across implementations. For example, JSON.stringify on the same object will always produce the same string, and JSON.parse(JSON.stringify(obj)) would produce an object with the same key ordering as the original (assuming the object is completely JSON-serializable).

kitop commented:
Right, so the same object will always produce the same string. But 2 different objects with the same key-values but different key ordering are not guaranteed to produce the same string, right?

So, having: `{ from: "/path", to: "/foo" }` and `{ to: "/foo", from: "/path" }` might produce different strings. Similar to what Lukas pointed out.

lukasholzer reviewed (changes requested):
I would use a different way for generating a key as a `JSON.stringify` out of the previously mentioned reasons.

Furthermore If we would convert the redirects parser first to vitest framework like the other packages you could leverage the benchmark feature of it to test the performance implications https://vitest.dev/guide/features.html#benchmarking-experimental

Should be a rather quick thing I might try to push a PR

lukasholzer commented on packages/redirect-parser/src/merge.ts:
```diff
@@ -10,14 +8,23 @@ import { splitResults } from './results.js'
+const removeDuplicates = function (redirects) {
+  const uniqueRedirects = new Set()
+  const result = []
+  redirects.reverse().forEach((r) => {
```
I would do something like this for creating the key:

```typescript
const r = redirects[i]
const key = createHash('md5').update(`${r.from}${r.to}${r.status}${r.force}${query}`).digest('hex');
```

JGAntunes commented:
Ah nice catch! Thanks y'all, I'll review these afterwards 👍

JGAntunes commented:
I've updated the PR to use a stable JSON stringify implementation (using a package we already use in other places within build - https://github.com/davidmarkclements/fast-safe-stringify). Also added a vitest `bench` set of tests.

> there are some libraries that provide deterministic stringify but I would think of a more performant solution. Create a function that hashes a redirect and you define the order on your own.

I think having an implementation that works independently of the redirect format is a safer bet. We would have to either define our own key format  (which we would need to change if ever came the time where the redirect object format changes) or have our own stable stringify approach (effectively what the module does for us), all for what IMO would be a really small performance improvement (which might not even be that much if we ended up hashing each key).

JGAntunes commented on packages/redirect-parser/src/merge.ts:
```diff
@@ -10,14 +10,24 @@ import { splitResults } from './results.js'
+    uniqueRedirects.add(key)
+    result.unshift(r)
+  }
+  return result
```
Good call 👍 , changed in https://github.com/netlify/build/pull/4652/commits/5e20137272a465cfe7ee77489e2322989f2c8bde

JGAntunes commented:
> This pull request adds or modifies JavaScript (.js, .cjs, .mjs) files.
Consider converting them to TypeScript.

Already done I believe (double checked but I don't think I missed anything) 🤔 maybe it's picking up the file renames too?
